Marlin/Cartridge.cpp

```diff
@@ -0,0 +1,237 @@
+    // mark it as removed.
+    if (cartridgesRemovedCheck()) {
+        cartridgeRemovalHysteresis = CARTRIDGE_REMOVAL_HYSTERESIS_COUNT;
+    } else if (removedCondition) {
+        cartridgeRemovalHysteresis = CARTRIDGE_REMOVAL_HYSTERESIS_COUNT;
+    }
+
+    // We have this hysteresis here to accomodate putting the cartridge
+    // back in. Without it, Marlin will recognize the cartridge has been
+    // reinserted before the temperature updates from its maximum value and
+    // will throw a high temperature error.
+    if (cartridgeRemovalHysteresis > 0) {
+        cartridgeRemovalHysteresis--;
+        returnValue = true;
+    }
+
+    return returnValue;
+}
+
+/**
+ * This function checks to see if the FFF cartridge is removed,
+ * to prevent heating
+ * @returns    Returns true if an FFF cartridge has been removed
+ */
+bool CartridgeRemovedFFF(void) {
+    if (!cartridgePresent[0]) {
+        return true;
+    }
+    return false;
+}
+
+/**
+ * This is the error handler for when we see the cartridge removed error. It 
+ * reports to octoprint that we should pause the print, as well as disabling
+ * all heaters and stopping the currently queued commands. This will only get
+ * called once if it's been called multiple times in quick succession.
+ * @inputs     An input that will be displayed on the serial monitor
+ */
+void _cartridge_removed_error(const char *serial_msg) {
+    if (IsSafetyCriticalSection()) {
+        static bool killed = false;
+        if (IsRunning()) {
+            SERIAL_ERROR_START;
+            serialprintPGM(serial_msg);
+            SERIAL_EOL;
+        }
+        if (!killed) {
+            Running = false;
+            killed = true;
+            kill(serial_msg);
+        }
+    } else {
+        static millis_t timeSinceLastRemoval = {0};
+        if (millis() > timeSinceLastRemoval + CARTRIDGE_REMOVED_ERR_INTERVAL) {
+            quickStop();
+            disable_all_heaters();
+            disable_all_steppers();
+            serialprintPGM(serial_msg);
+            SERIAL_EOL;
+            SERIAL_ECHOLN("// action:pause");
+        }
+        timeSinceLastRemoval = millis();
+    }
+}
+
+/**
+ * Macro function that updates cartridge status, and checks if a cartridge has 
+ * been removed.
+ * @returns    Returns true if a cartridge is removed
+ */
+bool CartridgeUpdateAndCheck() {
+    bool returnValue = false;
+    UpdateCartridgeStatus();
+    returnValue = CartridgeRemoved();
+    return returnValue;
+}
+
+//===========================================================================
+//============================ Private Functions ============================
+//===========================================================================
+
+/**
+ * Reports that a cartridge is absent. If there was a cartridge present,
+ * marks it as removed
+ */
+static void cartridgeAbsentUpdate(unsigned int cartNumber) {
+    if (cartridgePresent[cartNumber] == true) {
```

Comment:
is the `== true` part needed here? Couldn't it just be `if (cartridgePresent[cartNumber]) {` ?